### ISSUE TITLE
Use `displayException` instead of `show` in top-level exception handler

### ISSUE THREAD

**Issue by bgamari** [haskell/core-libraries-committee#198]
When GHC's extensible exception mechanism was introduced, it did not provide the `Exception` class with a method to render exceptions to a string. Instead, GHC's implementation opted to use the `show` function. This introduced an uneasy tension:

 * We want error messages being displayed to the user to be easy-to-understand
 * We want the `show` to produce valid Haskell syntax (being the inverse of `read`)

In 2008 this tension was [noticed](https://mail.haskell.org/pipermail/libraries/2014-November/024176.html) and [addressed](https://gitlab.haskell.org/ghc/ghc/-/issues/9822) with the introduction of the `displayException` method into `Exception`:
```haskell
class (Show a, Typeable a) => Exception a where
    ...
    displayException :: a -> Show
    displayException = show
```

However, at the time the CLC stopped short of changing the top-level exception handler (which handles exceptions not caught by the user) to use this function to display messages to the user.

Here I propose that we do precisely this. While this is a change in semantics, I argue that it is worthwhile as:

 * it only affects uncaught exceptions
 * it only affects output of `stderr`, which is already fragile at best
 * all exceptions defined by `base` (save `Deadlock`) use the default `show` implementation of `displayException` and therefore will not change in behavior

Furthermore, this change will facilitate showing of useful annotation information introduced by the [exception backtrace proposal](https://github.com/ghc-proposals/ghc-proposals/blob/b4b6042760444a5fdfb2f78d3ab01777c7cde8d4/proposals/0330-exception-backtraces.rst).

**Comment by phadej** [comment]
Is  top-level exception handler defined in `base`?

**Comment by bgamari** [comment]
> Is top-level exception handler defined in `base`?

It's defined in `GHC.TopHandler`. While I identified this being a potential "internal" module in #146, it nevertheless resides in `base` and the user-impact of the change IMHO warrants a CLC proposal.

**Comment by hasufell** [comment]
I'm not sure I even consider this a breaking change. 

PVP doesn't. People using error messages for control flow should get their source code deleted by GHC.

Test suites might break. But everyone who has such a test suite (like golden tests) is already used to such issues and should have mechanisms to fix it.

**Comment by bgamari** [comment]
> I'm not sure I even consider this a breaking change.

For what it's worth, I agree. However, I am intentionally being quite liberal in my definition of "breaking" here to avoid unnecessary impact. Exposing the proposed `oldDefaultHandler` is a relatively easy measure which shouldn't significantly affect maintainability. 

Of course, if the CLC believes that the proposal would be stronger without the measure then I am happy to remove it.

**Comment by Bodigrim** [comment]
Looks good to me. Dear CLC members, any more opinions on this before we vote?
@tomjaguarpaw @velveteer @angerman @mixphix @parsonsmatt

**Comment by parsonsmatt** [comment]
+1. I don't consider this a breaking change.

**Comment by velveteer** [comment]
+1

**Comment by mixphix** [comment]
+1

**Comment by Bodigrim** [comment]
Dear CLC members, let's vote on the proposal to use `displayException` instead of `show` in top-level exception handler as detailed above. 

@tomjaguarpaw @angerman @hasufell 

----

+1 from me.

**Comment by hasufell** [comment]
+1

**Comment by adamgundry** [comment]
@tomjaguarpaw The big improvement from this proposal comes from the fact that using `displayException` will add a backtrace (once #200 is implemented), whereas `show` will not. Do you have an alternative suggestion for how backtraces should be made visible in the top-level exception handler?

Your post sounds like a good case for having `displayException` implementations include searchable names (which they mostly do at the moment, at least in `base`, since apart from `Deadlock` they are basically the same as `show` anyway).

**Comment by tomjaguarpaw** [comment]
The vote has already passed, so at this point it's academic.

> in `base` ... apart from `Deadlock` they are basically the same as `show` anyway

Yes indeed, the ship has long sailed. My dissent is not really one that requires action from anyone, but I did think it was worth raising the issue.

> The big improvement from this proposal comes from the fact that using `displayException` will add a backtrace (once https://github.com/haskell/core-libraries-committee/issues/200 is implemented),

I see. I'm not really following the grand narrative of this exception redesign. That particular design choice sounds strange to me. Why allow display of backtraces to be overridden at the exception definition site? Anyway, I'm sure it will all become clear once we tackle that particular proposal.

**Comment by bgamari** [comment]
> Also, no biggie, but `Show` should be `String`:

Fixed.

**Comment by parsonsmatt** [comment]
@tomjaguarpaw That's a great point. What if we *also* added a `print $ typeOf exn` to the default exception handler?

**Comment by tomjaguarpaw** [comment]
I think that would probably completely resolve my concern, and even resolve the problem with the status quo, which is similar.

**Comment by tomjaguarpaw** [comment]
> If there is interest in `print $ typeOf exn` amendment, please raise a separate proposal.

Isn't that a GHC detail, rather than something that would appear in `base`?

**Comment by tomjaguarpaw** [comment]
Oh no, I see, it would be the same thing as the `defaultHandler` that appears in this proposal.

**Comment by bgamari** [comment]
Thank you all.

Printing the type of uncaught exceptions is an interesting possibility. @tomjaguarpaw would you be willing to open a follow-up proposal with this idea?

**Comment by tomjaguarpaw** [comment]
Yes, done at https://github.com/haskell/core-libraries-committee/issues/231, though if this passes I would prefer not to be involved in implementation since I am rather short of spare cycles right now.
